### PR TITLE
Do not throw an error when restarting instead of refreshing

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
[AB#4015](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4015)

When the summarizer is stopped because we want to restart it instead of refreshing from an ack, we throw an error to close the container and prevent a network call.

Reference PR: https://github.com/microsoft/FluidFramework/pull/15009

The goal of this PR is to 
- send a generic telemetry event without sending any error telemetry events when we decide to stop the summarizer instead of refreshing the ack.
- prevent the execution of making a network call to fetch the snapshot (I don't know if this is important or not)

If we don't implement this fix, we'll see these events, some as error events.

Spot checking a few docs, we would raise 5 error events each time we close the summarizer instead of refreshing.
```
fluid:telemetry:Container:ContainerClose | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
fluid:telemetry:SummarizerNode:refreshLatestSummary_cancel | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
fluid:telemetry:Summarizer:Running:HandleLastSummaryAckError | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
fluid:telemetry:Summarizer:Running:FailToSummarize | Restarting summarizer instead of refreshing | genericError | submitSummaryFailure: Error while generating, uploading, or submitting summary

```

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/summarizeRestart.spec.ts [hunk 1]
```diff
@@ -168,22 +152,14 @@ describeNoCompat("Summarizer closes instead of refreshing", (getTestObjectProvid
...
-			});
 
+			assert(!summarizingContainer2.closed, "Summarizer2 should not have closed yet!");
 			// This tells the summarizer to process the latest summary ack
 			// This is because the second summarizer is not the elected summarizer and thus the summaryManager does not
 			// tell the summarizer to process acks.
 			await summarizer2.run("test");
 
 			assert(summarizingContainer2.closed, "Unknown acks should close the summarizer");
-			assert.equal(
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/summarizeRestart.spec.ts:182]: Note with the change I made, I lose the ability to generate this error. Furthermore, I don't know if it actually stops the network call anymore.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -3181,11 +3181,16 @@ export class ContainerRuntime
...
+			assert(
+				this._summarizer !== undefined,
+				"Should be only summarizer refreshing from storage",
+			);
+			this.mc.logger.sendTelemetryEvent({
+				eventName: "RestartInsteadOfRefreshFromServerFetch",
+				message: "Restarting summarizer instead of refreshing",
+			});
+			this._summarizer.stop("latestSummaryStateStale");
+			this.closeFn();
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3193]: There's nothing stopping the network call from executing anymore.

**Comment by NicholasCouri** [comment]
> 			this.summarizerNode.startSummary(summaryRefSeqNum, summaryNumberLogger);

do we need to invoke this in case refreshLatestAckFailed ? 

Shuldn't we simply return  on line 2581-2 ?
return {
     stage: "base",
     referenceSequenceNumber: summaryRefSeqNum,
     minimumSequenceNumber,
     error: continueResult.error,
    };

---
Refers to: packages/runtime/container-runtime/src/containerRuntime.ts:2599 in 8e3c39c. [](commit_id = 8e3c39c0436bed33f5d9e5c531ee34b60aa708e6, deletion_comment = False)

**Comment by agarwal-navin** [comment]
I am not sure I understand what the issue is and how this change prevents it. What causes the cascade of errors? Is the goal to send a telemetry instead of throwing an error when restarting?
Can you add some more context in the description?

**Comment by tyler-cai-microsoft** [comment]
> I am not sure I understand what the issue is and how this change prevents it. What causes the cascade of errors? Is the goal to send a telemetry instead of throwing an error when restarting? Can you add some more context in the description?

updated

**Comment by NicholasCouri** [comment]
> > ```
> >   	this.summarizerNode.startSummary(summaryRefSeqNum, summaryNumberLogger);
> > ```
> 
> do we need to invoke this in case refreshLatestAckFailed ?
> 
> Shuldn't we simply return on line 2581-2 ? return { stage: "base", referenceSequenceNumber: summaryRefSeqNum, minimumSequenceNumber, error: continueResult.error, };
> 
> Or even invoke checkContinue there ?
> 
> Refers to: packages/runtime/container-runtime/src/containerRuntime.ts:2599 in [8e3c39c](https://github.com/microsoft/FluidFramework/commit/8e3c39c0436bed33f5d9e5c531ee34b60aa708e6). [](commit_id = [8e3c39c](https://github.com/microsoft/FluidFramework/commit/8e3c39c0436bed33f5d9e5c531ee34b60aa708e6), deletion_comment = False)

I sync up with Tyler and the container is closed already at this point - we should be good.

**Review by NicholasCouri — APPROVED** [review]

**Comment by agarwal-navin** [comment]
> If we don't implement this fix, we'll see these events, some as error events.
> 
> Spot checking a few docs, we would raise 5 error events each time we close the summarizer instead of refreshing.
> 
> ```
> fluid:telemetry:Container:ContainerClose | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> fluid:telemetry:SummarizerNode:refreshLatestSummary_cancel | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> fluid:telemetry:Summarizer:Running:HandleLastSummaryAckError | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> fluid:telemetry:Summarizer:Running:FailToSummarize | Restarting summarizer instead of refreshing | genericError | submitSummaryFailure: Error while generating, uploading, or submitting summary
> ```

You will see these 4 (not 5) errors during the second summary attempt. Bascially, there are 2 cases:
1. Second summary attempt [here](https://github.com/microsoft/FluidFramework/blob/7494374435eec323d0285f08063d7798a86dca86/packages/runtime/container-runtime/src/summary/runningSummarizer.ts#L603) - This will always fail with your change and the 4 errors will be seen. We should just not do this attempt if the feature flag for restarting is enabled. Instead, close the summarizer there itself rather than doing this when attempting to refresh from snapshot.
2. Refresh on getting an ack - This is the more interesting one and you will see the first 4 errors. You should remove the `HandleLastSummaryAckError` error altogether because it's a duplicate of refreshLatestSummary_cancel. That would leave only 2 errors which is fine IMO.

File: packages/runtime/container-runtime/src/summary/summaryGenerator.ts [hunk 3]
```diff
@@ -252,7 +253,9 @@ export class SummaryGenerator {
 			// If failure happened on upload, we may not yet realized that socket disconnected, so check
 			// offlineError too.
 			const category =
-				cancellationToken.cancelled || error?.errorType === DriverErrorType.offlineError
+				cancellationToken.cancelled ||
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summaryGenerator.ts:256]: As per my comment [here](https://github.com/microsoft/FluidFramework/pull/15036#issuecomment-1507303691), we should not do this. The second summary attempt will always fail so don't even do it.

File: packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts [hunk 4]
```diff
@@ -371,10 +372,19 @@ export class SummarizerNode implements IRootSummarizerNode {
 					return { latestSummaryUpdated: false };
 				}
 
+				let latestSnapshotFetch: IFetchSnapshotResult;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts:375]: This does not look right. It's more of a workaround. There are too many try-catch blocks here + container runtime which are trying to do similar things. A try-catch block around the outermost call in container runtime (this.summarizerNode.refreshLatestSummary) should be sufficient, no?

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 5]
```diff
@@ -656,7 +657,7 @@ export class RunningSummarizer implements IDisposable {
 					);
 					const result = await resultSummarize.receivedSummaryAckOrNack;
 
-					if (result.success) {
+					if (result.success || result.error === restartErrorMessage) {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:660]: As per my comment https://github.com/microsoft/FluidFramework/pull/15036#issuecomment-1507303691, we should not do this. The second summary attempt will always fail so don't even do it.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3246]: Originally, I was throwing an error here and letting the errors propagate through the system causing a spike in error telemetry logs

**Comment by NicholasCouri** [comment]
> > If we don't implement this fix, we'll see these events, some as error events.
> > Spot checking a few docs, we would raise 5 error events each time we close the summarizer instead of refreshing.
> > ```
> > fluid:telemetry:Container:ContainerClose | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:SummarizerNode:refreshLatestSummary_cancel | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:Summarizer:Running:HandleLastSummaryAckError | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:Summarizer:Running:FailToSummarize | Restarting summarizer instead of refreshing | genericError | submitSummaryFailure: Error while generating, uploading, or submitting summary
> > ```
> 
> You will see these 4 (not 5) errors during the second summary attempt. Bascially, there are 2 cases:
> 
> 1. Second summary attempt [here](https://github.com/microsoft/FluidFramework/blob/7494374435eec323d0285f08063d7798a86dca86/packages/runtime/container-runtime/src/summary/runningSummarizer.ts#L603) - This will always fail with your change and the 4 errors will be seen. We should just not do this attempt if the feature flag for restarting is enabled. Instead, close the summarizer there itself rather than doing this when attempting to refresh from snapshot.
> 2. Refresh on getting an ack - This is the more interesting one and you will see the first 4 errors. You should remove the `HandleLastSummaryAckError` error altogether because it's a duplicate of refreshLatestSummary_cancel. That would leave only 2 errors which is fine IMO.

I actually like it !

**Comment by agarwal-navin** [comment]
> > If we don't implement this fix, we'll see these events, some as error events.
> > Spot checking a few docs, we would raise 5 error events each time we close the summarizer instead of refreshing.
> > ```
> > fluid:telemetry:Container:ContainerClose | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:SummarizerNode:refreshLatestSummary_cancel | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:Summarizer:Running:HandleLastSummaryAckError | Restarting summarizer instead of refreshing | genericError | Restarting summarizer instead of refreshing
> > fluid:telemetry:Summarizer:Running:FailToSummarize | Restarting summarizer instead of refreshing | genericError | submitSummaryFailure: Error while generating, uploading, or submitting summary
> > ```
> 
> You will see these 4 (not 5) errors during the second summary attempt. Bascially, there are 2 cases:
> 
> 1. Second summary attempt [here](https://github.com/microsoft/FluidFramework/blob/7494374435eec323d0285f08063d7798a86dca86/packages/runtime/container-runtime/src/summary/runningSummarizer.ts#L603) - This will always fail with your change and the 4 errors will be seen. We should just not do this attempt if the feature flag for restarting is enabled. Instead, close the summarizer there itself rather than doing this when attempting to refresh from snapshot.
> 2. Refresh on getting an ack - This is the more interesting one and you will see the first 4 errors. You should remove the `HandleLastSummaryAckError` error altogether because it's a duplicate of refreshLatestSummary_cancel. That would leave only 2 errors which is fine IMO.

In case it wasn't clear, with these, you don't need any of the changes in container runtime / summarizer node. Basically, the errors thrown in refresh code is just fine.

**Pull request closed without merging**